chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,66 +39,174 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
+        if (free_list.empty()) {
+            continue;
+        }
+        std::unique_ptr<Chunk> chunk = std::move(free_list.back());
+        CHAINERX_ASSERT(chunk != nullptr);
+        free_list.pop_back();
+        if (distance > kCompactionThreashold) {
+            CompactFreeBins(it_start, it);
+        }
+        return chunk;
+    }
+
+    return nullptr;
+}
+
+// Removes a chunk from an appropriate free list, and returns the removed chunk
+//
+// Not thread-safe
+std::unique_ptr<Chunk> MemoryPool::RemoveChunkFromFreeList(Chunk* chunk) {
+    CHAINERX_ASSERT(chunk != nullptr);
+
+    // Find an appropriate free list
+    auto free_bins_it = free_bins_.find(chunk->bytesize());
+    if (free_bins_it == free_bins_.end()) {
+        return nullptr;
+    }
+    FreeList& free_list = free_bins_it->second;
+
+    // Remove the given chunk from the found free list
+    auto it = std::find_if(free_list.begin(), free_list.end(), [chunk](const auto& ptr) { return ptr.get() == chunk; });
+    if (it == free_list.end()) return nullptr;
+    std::unique_ptr<Chunk> chunk_uniq_ptr = std::move(*it);
+    CHAINERX_ASSERT(chunk_uniq_ptr != nullptr);
+    free_list.erase(it);
+    return chunk_uniq_ptr;
+}
+
 MemoryPool::~MemoryPool() {
     // NOTE: CudaSetDeviceScope is not available at dtor because it may throw
     int orig_device_index{0};
     cudaGetDevice(&orig_device_index);
     cudaSetDevice(device_index_);
 
-    for (const std::vector<void*>& free_list : free_bins_) {
-        for (void* ptr : free_list) {
-            allocator_->Free(ptr);
+    for (auto& item : free_bins_) {
+        FreeList& free_list = item.second;
+        for (const std::unique_ptr<Chunk>& chunk : free_list) {
+            if (chunk->prev() == nullptr) {
+                allocator_->Free(chunk->ptr());
+            }
         }
     }
     // Ideally, in_use_ should be empty, but it could happen that shared ptrs to memories allocated
     // by this memory pool are released after this memory pool is destructed.
     // Our approach is that we anyway free CUDA memories held by this memory pool here in such case.
     // Operators of arrays holding such memories will be broken, but are not supported.
     for (const auto& item : in_use_) {
-        void* ptr = item.first;
-        allocator_->Free(ptr);
+        const std::unique_ptr<Chunk>& chunk = item.second;
+        if (chunk->prev() == nullptr) {
+            allocator_->Free(chunk->ptr());
+        }
     }
 
     cudaSetDevice(orig_device_index);
 }
 
 void MemoryPool::FreeUnusedBlocks() {
     CudaSetDeviceScope scope{device_index_};
-    std::lock_guard<std::mutex> lock{free_bins_mutex_};
-    for (const std::vector<void*>& free_list : free_bins_) {
-        for (void* ptr : free_list) {
-            allocator_->Free(ptr);
+
+    // Frees unused memory blocks
+    for (auto& pair : free_bins_) {
+        FreeList& free_list = pair.second;
+        FreeList new_free_list;
```

Comment:
This requires extra malloc for the new vector. I don't think that's worthwhile.
Rather than that, how about this logic:
```
for (auto& chunk: free_list) {
    if (...) {
        allocator_->Free(chunk->ptr());
        chunk.reset();
    }
}
free_list.erase(
    std::remove_if(free_list.begin(), free_list.end(), [](const auto& chunk) { return chunk == nullptr; }),
    free_list.end());
```